extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../../../src/3p-analytics-common';
+import {throttle} from '../../../src/utils/rate-limit';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.CrossDomainIframeMessageQueue';
+
+/** @const {number} */
+export const MESSAGE_THROTTLE_TIME = 100;
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ * @abstract
+ */
+class AbstractAmpAnalytics3pMessageQueue {
```

Comment:
We plan to handle the batching use case in a different level. @zhouyx has started looking into it.
 
Ideally, pub should be able to do 
```js
transport: {
  beacon: true,
  xhrpost: true,
  image: true,
  batchRequestsInMs: 100,
}
```
For example, we can enable it for the GoogleAnalytics post API, which does support batching. 

So can you hold the message queue code for now, and assume throttling will happen in a later phase.